File: backend/src/main/java/com/radar/prospection/api/RadarController.java
package com.radar.prospection.api;

import com.radar.prospection.domain.*;
import com.radar.prospection.repository.MissionRepository;
import com.radar.prospection.repository.SignalRepository;
import com.radar.prospection.scraper.ScraperCoordinator;
import com.radar.prospection.signal.SignalCoordinator;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/radar")
@CrossOrigin(origins = "*")
@RequiredArgsConstructor
public class RadarController {

    private final MissionRepository missionRepository;
    private final SignalRepository signalRepository;
    private final ScraperCoordinator scraperCoordinator;
    private final SignalCoordinator signalCoordinator;

    // ──────────────────────────────────────────
    // MISSIONS
    // ──────────────────────────────────────────

    @GetMapping("/missions")
    public List<MissionDto> getMissions(
            @RequestParam(defaultValue = "0") int minScore,
            @RequestParam(required = false) String status) {

        List<Mission> missions = minScore > 0
                ? missionRepository.findByFitScoreGreaterThanEqualOrderByFitScoreDesc(minScore)
                : status != null
                    ? missionRepository.findByStatusOrderByFitScoreDesc(MissionStatus.valueOf(status))
                    : missionRepository.findRecentMissions(LocalDateTime.now().minusDays(7));

        return missions.stream().map(MissionDto::from).toList();
    }

    @GetMapping("/missions/top")
    public List<MissionDto> getTopMissions() {
        return missionRepository.findByFitScoreGreaterThanEqualOrderByFitScoreDesc(70)
                .stream().map(MissionDto::from).toList();
    }

    @PatchMapping("/missions/{id}/status")
    public ResponseEntity<MissionDto> updateStatus(
            @PathVariable Long id,
            @RequestBody Map<String, String> body) {

        return missionRepository.findById(id).map(mission -> {
            mission.setStatus(MissionStatus.valueOf(body.get("status")));
            return ResponseEntity.ok(MissionDto.from(missionRepository.save(mission)));
        }).orElse(ResponseEntity.notFound().build());
    }

    // ──────────────────────────────────────────
    // SIGNAUX
    // ──────────────────────────────────────────

    @GetMapping("/signals")
    public List<SignalDto> getSignals(
            @RequestParam(defaultValue = "0") int minScore) {

        return (minScore > 0
                ? signalRepository.findHighValueSignals(minScore)
                : signalRepository.findByStatusOrderByOpportunityScoreDesc(SignalStatus.NEW))
                .stream().map(SignalDto::from).toList();
    }

    @GetMapping("/signals/hot")
    public List<SignalDto> getHotSignals() {
        return signalRepository.findHighValueSignals(60)
                .stream().map(SignalDto::from).toList();
    }

    // ──────────────────────────────────────────
    // ACTIONS MANUELLES
    // ──────────────────────────────────────────

    @PostMapping("/scan/missions")
    public ResponseEntity<Map<String, Object>> triggerMissionScan() {
        ScraperCoordinator.ScrapingResult result = scraperCoordinator.runAll();
        return ResponseEntity.ok(Map.of(
            "newMissions", result.newMissions(),
            "duplicates", result.duplicates(),
            "analyzed", result.analyzed(),
            "errors", result.errors()
        ));
    }

    @PostMapping("/scan/signals")
    public ResponseEntity<Map<String, Object>> triggerSignalScan() {
        SignalCoordinator.SignalScanResult result = signalCoordinator.runAll();
        return ResponseEntity.ok(Map.of(
            "newSignals", result.newSignals(),
            "errors", result.errors()
        ));
    }

    // ──────────────────────────────────────────
    // DASHBOARD STATS
    // ──────────────────────────────────────────

    @GetMapping("/stats")
    public Map<String, Object> getStats() {
        LocalDateTime since = LocalDateTime.now().minusDays(7);
        List<Mission> recent = missionRepository.findRecentMissions(since);
        List<Signal> hotSignals = signalRepository.findHighValueSignals(60);

        long topMissions = recent.stream().filter(m -> m.getFitScore() != null && m.getFitScore() >= 70).count();
        double avgScore = recent.stream()
                .filter(m -> m.getFitScore() != null)
                .mapToInt(Mission::getFitScore)
                .average().orElse(0);

        return Map.of(
            "missionsLast7Days", recent.size(),
            "topMissions", topMissions,
            "avgFitScore", Math.round(avgScore),
            "hotSignals", hotSignals.size(),
            "missionsBySource", countBySource(recent)
        );
    }

    private Map<String, Long> countBySource(List<Mission> missions) {
        Map<String, Long> counts = new java.util.LinkedHashMap<>();
        for (Mission m : missions) {
            counts.merge(m.getSource().name(), 1L, Long::sum);
        }
        return counts;
    }
}

File: backend/src/main/kotlin/com/radar/prospection/api/RadarController.kt
package com.radar.prospection.api

import com.radar.prospection.claude.MissionAnalysisService
import com.radar.prospection.claude.OutreachService
import com.radar.prospection.domain.MissionStatus
import com.radar.prospection.domain.SignalStatus
import com.radar.prospection.repository.MissionRepository
import com.radar.prospection.repository.SignalRepository
import com.radar.prospection.scraper.ScraperCoordinator
import com.radar.prospection.signal.SignalCoordinator
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.time.LocalDateTime

@RestController
@RequestMapping("/api/radar")
class RadarController(
    private val missionRepository: MissionRepository,
    private val signalRepository: SignalRepository,
    private val scraperCoordinator: ScraperCoordinator,
    private val signalCoordinator: SignalCoordinator,
    private val analysisService: MissionAnalysisService,
    private val outreachService: OutreachService
) {

    @GetMapping("/missions")
    fun getMissions(
        @RequestParam(defaultValue = "0") minScore: Int,
        @RequestParam(required = false) status: String?
    ): List<MissionDto> {
        val missions = when {
            minScore > 0 -> missionRepository.findByFitScoreGreaterThanEqualOrderByFitScoreDesc(minScore)
            status != null -> missionRepository.findByStatusOrderByFitScoreDesc(MissionStatus.valueOf(status))
            else -> missionRepository.findRecentMissions(LocalDateTime.now().minusDays(7))
        }
        return missions.map { MissionDto.from(it) }
    }

    @GetMapping("/missions/top")
    fun getTopMissions(): List<MissionDto> =
        missionRepository.findByFitScoreGreaterThanEqualOrderByFitScoreDesc(70).map { MissionDto.from(it) }

    @PatchMapping("/missions/{id}/status")
    fun updateStatus(@PathVariable id: Long, @RequestBody body: Map<String, String>): ResponseEntity<MissionDto> =
        missionRepository.findById(id).map { mission ->
            mission.status = MissionStatus.valueOf(body["status"]!!)
            ResponseEntity.ok(MissionDto.from(missionRepository.save(mission)))
        }.orElse(ResponseEntity.notFound().build())

    @GetMapping("/signals")
    fun getSignals(@RequestParam(defaultValue = "0") minScore: Int): List<SignalDto> {
        val signals = if (minScore > 0)
            signalRepository.findHighValueSignals(minScore)
        else
            signalRepository.findByStatusOrderByOpportunityScoreDesc(SignalStatus.NEW)
        return signals.map { SignalDto.from(it) }
    }

    @GetMapping("/signals/hot")
    fun getHotSignals(): List<SignalDto> =
        signalRepository.findHighValueSignals(60).map { SignalDto.from(it) }

    @PostMapping("/scan/missions")
    fun triggerMissionScan(): ResponseEntity<Map<String, Any>> {
        val result = scraperCoordinator.runAll()
        return ResponseEntity.ok(mapOf(
            "newMissions" to result.newMissions,
            "duplicates" to result.duplicates,
            "analyzed" to result.analyzed,
            "errors" to result.errors
        ))
    }

    @PostMapping("/reanalyze")
    fun reanalyzeMissions(): ResponseEntity<Map<String, Any>> {
        val pending = missionRepository.findByFitScoreIsNull()
        var count = 0
        for (m in pending) {
            try {
                missionRepository.save(analysisService.analyze(m))
                count++
            } catch (_: Exception) {}
        }
        return ResponseEntity.ok(mapOf("reanalyzed" to count, "total" to pending.size))
    }

    @GetMapping("/missions/favorites")
    fun getFavorites(): List<MissionDto> =
        missionRepository.findByFavoriteTrueOrderByFitScoreDesc().map { MissionDto.from(it) }

    @PatchMapping("/missions/{id}/favorite")
    fun toggleFavorite(@PathVariable id: Long, @RequestBody body: Map<String, Boolean>): ResponseEntity<MissionDto> =
        missionRepository.findById(id).map { mission ->
            mission.favorite = body["favorite"] ?: false
            ResponseEntity.ok(MissionDto.from(missionRepository.save(mission)))
        }.orElse(ResponseEntity.notFound().build())

    @PostMapping("/missions/{id}/outreach")
    fun generateOutreach(@PathVariable id: Long): ResponseEntity<OutreachDto> =
        missionRepository.findById(id)
            .map { ResponseEntity.ok(outreachService.generate(it)) }
            .orElse(ResponseEntity.notFound().build())

    @PostMapping("/scan/signals")
    fun triggerSignalScan(): ResponseEntity<Map<String, Any>> {
        val result = signalCoordinator.runAll()
        return ResponseEntity.ok(mapOf(
            "newSignals" to result.newSignals,
            "errors" to result.errors
        ))
    }

    @GetMapping("/stats")
    fun getStats(): Map<String, Any> {
        val since = LocalDateTime.now().minusDays(7)
        val recent = missionRepository.findRecentMissions(since)
        val hotSignals = signalRepository.findHighValueSignals(60)

        val topMissions = recent.count { it.fitScore != null && it.fitScore!! >= 70 }
        val avgScore = recent.filter { it.fitScore != null }.map { it.fitScore!! }.average().let {
            if (it.isNaN()) 0.0 else it
        }

        return mapOf(
            "missionsLast7Days" to recent.size,
            "topMissions" to topMissions,
            "avgFitScore" to Math.round(avgScore),
            "hotSignals" to hotSignals.size,
            "missionsBySource" to recent.groupingBy { it.source?.name ?: "UNKNOWN" }.eachCount()
        )
    }
}
